### PR TITLE
[FLINK-17435] [hive] Hive non-partitioned source supports streaming read

### PR DESCRIPTION
**Pull request by godfreyhe**
## What is the purpose of the change

*for non-partitioned hive source, we can use existing continuous file processing mechanism to read new data. This pr aim let non-partitioned hive source support streaming read.*

## Brief change log
  - *A hive FileInputFormat that wraps HiveTableInputFormat*
  - *create streaming source with PROCESS_CONTINUOUSLY mode*

## Verifying this change

This change added tests and can be verified as follows:

  - *Extended HiveTableSourceTest for streaming read on non-partitioned source*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / docs / **JavaDocs** / not documented)

### CODE REVIEW

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/ConsumeOrder.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+	 */
+	CREATE_TIME_ORDER("create-time"),
+
+	/**
+	 * partition-time compare time represented by partition name.
+	 */
+	PARTITION_TIME_ORDER("partition-time");
+
+	private final String order;
+	ConsumeOrder(String order) {
```
Comment (JingsongLi) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/ConsumeOrder.java:39]: Above should have a blank line.

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/ConsumeOrder.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+	@Override
+	public String toString() {
+		return order;
+	}
+
+	/**
+	 * Get {@link ConsumeOrder} from consume order string.
+	 */
+	public static ConsumeOrder getConsumeOrder(String consumeOrderStr) {
+		for (ConsumeOrder consumeOrder : ConsumeOrder.values()) {
```
Comment (JingsongLi) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/ConsumeOrder.java:52]: `ConsumeOrder.values()` -> `values()`

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/ConsumeOrder.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+			if (consumeOrder.order.equals(consumeOrderStr)) {
```
Comment (JingsongLi) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/ConsumeOrder.java:53]: `equalsIgnoreCase`

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/ConsumeOrder.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+				return consumeOrder;
+			}
+		}
+		return null;
```
Comment (JingsongLi) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/ConsumeOrder.java:57]: throw exception, we should not pass a null to operators.

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/HiveTableSource.java [hunk 5]
```diff
@@ -261,6 +267,57 @@ private boolean isStreamingSource() {
...
+			StreamExecutionEnvironment execEnv,
+			TypeInformation<RowData> typeInfo,
+			HiveTableInputFormat inputFormat,
+			HiveTablePartition hiveTable) {
+		HiveTableFileInputFormat fileInputFormat = new HiveTableFileInputFormat(
+				inputFormat,
+				hiveTable);
+		fileInputFormat.setFilePath(getFilePath());
+
+		final Map<String, String> properties = catalogTable.getOptions();
```
Comment (JingsongLi) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/HiveTableSource.java:280]:
```
Configuration configuration = new Configuration();
options.forEach(configuration::setString)
```
We can use `Configuration` to get option.

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/HiveTableSource.java [hunk 6]
```diff
@@ -261,6 +267,57 @@ private boolean isStreamingSource() {
...
+		ConsumeOrder consumeOrder = ConsumeOrder.getConsumeOrder(consumeOrderStr);
+		if (consumeOrder != ConsumeOrder.CREATE_TIME_ORDER) {
+			throw new UnsupportedOperationException("Unsupported consumer order: " + consumeOrder);
+		}
+
+		String consumeOffset = properties.getOrDefault(
+				STREAMING_SOURCE_CONSUME_START_OFFSET.key(),
+				STREAMING_SOURCE_CONSUME_START_OFFSET.defaultValue());
+		long currentReadTime = Long.MIN_VALUE;
+		if (consumeOffset != null) {
```
Comment (JingsongLi) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/HiveTableSource.java:294]: never null.

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/HiveTableSource.java [hunk 7]
```diff
@@ -261,6 +267,57 @@ private boolean isStreamingSource() {
...
 
+	private DataStream<RowData> createStreamSourceForNonPartitionTable(
+			StreamExecutionEnvironment execEnv,
+			TypeInformation<RowData> typeInfo,
+			HiveTableInputFormat inputFormat,
+			HiveTablePartition hiveTable) {
+		HiveTableFileInputFormat fileInputFormat = new HiveTableFileInputFormat(
+				inputFormat,
+				hiveTable);
+		fileInputFormat.setFilePath(getFilePath());
```
Comment (JingsongLi) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/HiveTableSource.java:278]: Remove `getFilePath`, we can get path from `allHivePartitions`(should be a single object when no partitions.).

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/read/HiveTableFileInputFormat.java [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+import org.apache.flink.table.data.RowData;
+
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.mapred.FileSplit;
+
+import java.io.IOException;
+import java.net.URI;
+
+/**
+ * A {@link FileInputFormat} that wraps a {@link HiveTableInputFormat}.
```
Comment (JingsongLi) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/read/HiveTableFileInputFormat.java:35]: Please add comments:
- Only support FileInputSplit
- Only support renaming inserting 
- getSplit use flinks instead format.

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/read/HiveTableFileInputFormat.java [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+	}
+
+	@Override
+	public RowData nextRecord(RowData reuse) throws IOException {
+		return inputFormat.nextRecord(reuse);
+	}
+
+	@Override
+	public void configure(Configuration parameters) {
+		inputFormat.configure(parameters);
```
Comment (JingsongLi) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/read/HiveTableFileInputFormat.java:73]: A better way is first `super.`, second do own works.
same below.

File: flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/HiveTableSourceTest.java [hunk 10]
```diff
@@ -531,6 +534,76 @@ public void testStreamPartitionRead() throws Exception {
...
+	}
+
+	private void testNonPartitionStreamingSource(Boolean useMapredReader, String tblName) throws Exception {
+		final String catalogName = "hive";
+		final String dbName = "source_db";
+		hiveShell.execute("CREATE TABLE source_db." + tblName + " (" +
+				"  a INT," +
+				"  b CHAR(1) " +
+				") TBLPROPERTIES (" +
+				"  'streaming-source.enable'='true'," +
```
Comment (JingsongLi) [inline flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/HiveTableSourceTest.java:554]: `useMapredReader` only works in parquet and orc, you should use them.

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/read/HiveTableFileInputFormat.java (hunk 8 above)
Comment (godfreyhe) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/read/HiveTableFileInputFormat.java:35]: "Only support renaming inserting" is not the limitation of `HiveTableFileInputFormat`, it's the limit of `ContinuousFileMonitoringFunction`, I have add some documents in `FileSystemOptions`.

**Comment by rmetzger** [comment]
Note: This change broke master: https://dev.azure.com/apache-flink/apache-flink/_build/results?buildId=1429&view=logs&s=ae4f8708-9994-57d3-c2d7-b892156e7812&j=52b61abe-a3cc-5bde-cc35-1bbe89bb7df5

**Comment by rmetzger** [comment]
I reverted this change in https://github.com/apache/flink/commit/e69d21b6cd99201dfe4ae19d477064be7bc80be0

**Comment by rmetzger** [comment]
I reopened the ticket

**Comment by JingsongLi** [comment]
Thanks @rmetzger for fixing.

**Pull request merged**